src/persist-client/src/internal/metrics.rs

```diff
@@ -2417,6 +2422,177 @@ impl BlobMemCache {
     }
 }
 
+#[derive(Debug)]
+pub struct SemaphoreMetrics {
+    cfg: PersistConfig,
+    registry: MetricsRegistry,
+    fetch: OnceCell<MetricsSemaphore>,
+}
+
+impl SemaphoreMetrics {
+    fn new(cfg: PersistConfig, registry: MetricsRegistry) -> Self {
+        SemaphoreMetrics {
+            cfg,
+            registry,
+            fetch: OnceCell::new(),
+        }
+    }
+
+    /// We can't easily change the number of permits, and the dyncfgs are all
+    /// set to defaults on process start, so make sure we only initialize the
+    /// semaphore once we've synced dyncfgs at least once.
+    async fn fetch(&self) -> &MetricsSemaphore {
+        if let Some(x) = self.fetch.get() {
+            // Common case of already initialized avoids the cloning below.
+            return x;
+        }
+        let cfg = self.cfg.clone();
+        let registry = self.registry.clone();
+        let init = async move {
+            let total_permits = match cfg.announce_memory_limit {
+                // Non-cc replicas have the old physical flow control mechanism,
+                // so only apply this one on cc replicas.
+                Some(mem) if cfg.is_cc_active => {
+                    // We can't easily adjust the number of permits later, so
+                    // make sure we've synced dyncfg values at least once.
+                    info!("fetch semaphore awaiting first dyncfg values");
+                    let () = cfg.configs_synced_once().await;
+                    let total_permits = usize::cast_lossy(
+                        f64::cast_lossy(mem) * FETCH_SEMAPHORE_PERMIT_ADJUSTMENT.get(&cfg),
+                    );
+                    info!("fetch_semaphore got first dyncfg values");
+                    total_permits
+                }
+                Some(_) | None => Semaphore::MAX_PERMITS,
+            };
+            MetricsSemaphore::new(&registry, "fetch", total_permits)
+        };
+        self.fetch.get_or_init(|| init).await
+    }
+
+    pub(crate) async fn acquire_fetch_permits(&self, encoded_size_bytes: usize) -> MetricsPermits {
```

Comment:
Will this get stuck if we request more permits than the total number of permits? (If so, seems safish to cap `requested_permits` at `total_permits`.)

Reply:
It does! Just happens in `acquire_permits`

Reply:
🤦!!